### ISSUE TITLE
Faster API with fragment caching?

### ISSUE THREAD

**Issue by slaskis** [JSONAPI-Resources/jsonapi-resources#297]
I've previously tried to speed things up with eager loading of the includes. And it seems to work pretty well so far. And by [patching the JSON renderer](https://github.com/GoodLife/rails-patch-json-encode) the view rendering has gone from like 20% of the profiling time to less than the threshold.

But now I've narrowed the biggest bottleneck down to "building the JSON". Basically 90% of the requests are spent in the recursive `JSONAPI::ResourceSerializer#object_hash` and related methods. 

And this sometimes leaves a request log like this:

```
Completed 200 OK in 12242ms (Views: 278.9ms | ActiveRecord: 860.4ms)
```

Where it takes _860ms_ in the DB, _278ms_ rendering the JSON and _11104ms_ in "other", where my profiling points to the `object_hash`-methods mentioned above being almost all of the time spent.

Granted the queries get quite large in these rather complicated `?include`s but this particular response wasn't even bigger than ~400KB so it still seems a bit too much with 12s. And most of the times these requests are caught in view caches and varnish so it doesn't really break our app yet. But I still would like to keep the uncached requests at most 1 second.

So considering this I was thinking maybe it could be quite "easy" to implement [Rails fragment caching](http://guides.rubyonrails.org/caching_with_rails.html#fragment-caching)? 
I haven't used it myself yet in a project (never had a reason too) but it seems to be real easy if you use something like jbuilder for generating your JSON. 

In theory I'm naively thinking that it would be enough to just wrap [these lines](https://github.com/cerebris/jsonapi-resources/blob/master/lib/jsonapi/resource_serializer.rb#L101-L119) in a `cache(source.cache_key)`-block. 

Anyone with some insight into what else could go wrong with this approach?

**Comment by lgebhardt** [comment]
@slaskis Sorry for the long delay in looking at this. I've only briefly investigated this, so this is certainly not definitive.  While in a request (that has includes) there is a lot of time spent in the `object_hash` method, most of that time is spent in the `ActiveRecord::Associations::CollectionAssociation#reader` method. This is the rails code that gets and creates the (hopefully) eager loaded records.

We've recently fixed some bugs related to eager loading included resources, and not including ids if a resource is not included. Because of those two changes it's possible the code is now faster (by not loading has many related resources) then when you reported this issue. If you have some time would you mind looking at this again and providing an update?

**Comment by slaskis** [comment]
Great! 

Unfortunately I'm still quite a bit behind on that project, it's still on a fork of 0.3.3 which is before the finished spec so it requires updating a few consumers too. So it's hard to try that out in production at this moment. 

But I have tried making a perf-branch for the peeps example app and it shows a slight decrease of object allocations, not a huge difference (~4%) but still an improvement. But that's just the simplest example of 10 contacts x 100 phone numbers.

**Comment by johantell** [comment]
Has there been any progress on caching so far? I tried it out myself and had some decent improvements in performance but there were a lot of tests failing cause of it

**Comment by vrybas** [comment]
related issue #74

**Comment by DavidMikeSimon** [comment]
See merged PR #712
